### ISSUE TITLE
conditional/limited cobegin statement

### ISSUE THREAD

**Issue by cassella** [chapel-lang/chapel#13518]
`_MergeSort()` has this construct

```chapel
    if(here.runningTasks() < here.numPUs(logical=true)) {
      cobegin {
        { _MergeSort(Data, lo, mid, minlen, comparator); }
        { _MergeSort(Data, mid+stride, hi, minlen, comparator); }
      }
    } else {
      _MergeSort(Data, lo, mid, minlen, comparator);
      _MergeSort(Data, mid+stride, hi, minlen, comparator);
    }
```

This led me to wonder if there should be a way to make `cobegin`'s
parallelism conditional or limited by an argument, like the `serial`
or `local` statements' conditional,

```chapel
  cobegin (here.runningTasks() < here.numPUs(logical=true)) {
    { _MergeSort(...); }
    { _MergeSort(...); }
  }
```

Or provide directly the max number of tasks to create,

```chapel
  cobegin (here.numPUs(logical=true) - here.runningTasks()) {
    { _MergeSort(...); }
    { _MergeSort(...); }
  }
```

If there are more statements in the `cobegin` than tasks allowed, the
statements would be divided between the tasks, either statically or
dynamically.

Since the task entering the `cobegin` will be idle while the
statements are being executed, it should work out to not "counting"
against the limit.  So for `cobegin (1) {`, the two `_MergeSort()`
calls should be executed in parallel.

Or as above, but the expression is re-evaluated each time one of the
statements completes.  Or even whenever `here.runningTasks()` changes?

Finally, it would be nice if there were a shorthand way of writing the natural way to make use of this, whatever that turns out to be,

```chapel
  cobegin? {
    { _MergeSort(...); }
    { _MergeSort(...); }
  }
```

I thought about

```chapel
  serial (here.runningTasks >= here.numPUs(logical=true)) {
    cobegin {
      { _MergeSort(...); }
      { _MergeSort(...); }
    }
  }
```

But the task then can't "escape" from the serial constraint, even if
there are fewer running tasks later, when one of those calls to
`_MergeSort()` reaches the `cobegin`.

**Comment by cassella** [comment]
PR #14204 adds a different motivating example,

```chapel
    if hi-lo < 300 {
      // stay sequential
      quickSort(Data, minlen, comparator, lo, eqStart-1);
      quickSort(Data, minlen, comparator, eqEnd+1, hi);
    } else {
      // do the subproblems in parallel
      forall i in 1..2 {
        if i == 1 then
          quickSort(Data, minlen, comparator, lo, eqStart-1);
        else
          quickSort(Data, minlen, comparator, eqEnd+1, hi);
      }
    }
```

which imposes on the parallelism both a running-tasks constraint and a constraint unrelated to the number of running tasks, suggesting that the explicit way of writing a "conditional/limited cobegin" does need to maintain the "conditional" part, not just the "limited" part.

(Though this use would be satisfied by a conditional `serial` statement around a limited cobegin (or around its own `forall`, I think).  I'm not sure if that would be satisfactory generally, if the statements within the cobegin won't necessarily want to apply the same condition.)

It also suggests the parallelism can be described more succinctly than I managed above: A limited cobegin of `n` statements can be considered to be run in a `forall` loop over `1..n`, one statement per iteration.  Though the above example having exactly two statements may be too simple of an example wrt. load imbalance.  Perhaps there should be a way to specify a different iterator (or iterator factory function) to run over `1..n`, such as one of the `DynamicIters`?  (Of course, letting the compiler count the statements...)


```chapel
cobegin (hi - lo < 300, iter=DynamicIters.dynamic) {
    quickSort(Data, minlen, comparator, lo, eqStart-1);
    quickSort(Data, minlen, comparator, eqEnd+1, hi);
}
```

(Or `serial (hi - lo < 300) do cobegin (...`}

**Comment by mppf** [comment]
@cassella - I don't know if there is an issue about it but I have in the past advocated for having a variant of `begin` that doesn't create tasks if there are enough tasks going. (In particular I would expect the implementation to stop creating these tasks once the task queue is pretty full; it does that now for `forall`s generally but with a sortof parallel manual thing in the iterators (using `_computeNumChunks` which uses `here.runningTasks()`). I really like the `begin?` / `cobegin?` / `coforall?` syntax for that, since `?` already means "option" in the context of nilability. So thanks for the suggestion :)

**Comment by mppf** [comment]
Another idea here is to use `cobegin if (something)` to create parallelism only if `something` is `true`. This kindof reasoning can be extended to `begin` as well e.g. `begin if (something)`.

(With the original proposal, it could also be extended to `begin` as `begin (something)` ).

My own preference is to include the `if` in the construct, if we add it.

**Comment by bradcray** [comment]
@mppf: In your propoeal, how it seems it could be challenging to distinguish between a `begin` that wants to start a task that happens to be an `if`-statement vs. one that is using the `if` to indicate whether or not a task should be started?

**Comment by mppf** [comment]
@bradcray - that's a good point. We could use syntax like `begin (if something) f()`. That would be just as much a breaking change as `begin (something) f()` would. I think both are OK because `(something) f()` does not already parse. Of course, another approach would be to put it in the `with (...)` clause.
